### ISSUE TITLE
Poorer performance for GraphConv with uncertainty=True vs. False 

### ISSUE THREAD

**Issue by amandajean119** [deepchem/deepchem#1378]
**Description:**
A GraphConv model trained using the modified loss function with uncertainty=True shows consistently poorer R^2 scores and higher RMS scores than one trained using the original loss function.
On example, on Delaney:
Deepchem 2.0’s Graph convolution model,  250 epochs, 0.0005 learning rate, default layers and dropouts
•	Test-pearson_r2_score: 0.865,
•	Train-pearson_r2_score: 0.993,
•	Valid-pearson_r2_score: 0.859,
•	Test-rms_score: 0.808,
•	Train-rms_score: 0.288,
•	Valid-rms_score: 0.859
Deepchem 2.1’s GC with uncertainty, 250 epochs, 0.0005 learning rate, 0.6 dropout probability for all 50 masks
•	Test-pearson_r2_score: 0.812,
•	Train-pearson_r2_score: 0.804,
•	Valid-pearson_r2_score: 0.795,
•	Test-rms_score: 2.404,
•	Train-rms_score: 2.356,
•	Valid-rms_score: 2.388

This was also true when run multiple times:
![image](https://user-images.githubusercontent.com/3688087/44225309-021a6d80-a142-11e8-968d-77259bb54a86.png)
and true for both the Delaney dataset and the Tox21 dataset.

**Comment by peastman** [comment]
You're comparing a model trained with dropout to one trained without.  Those will of course be different.  Using dropout reduces the capacity of the model, and also may require longer training.  You're also using different versions of DeepChem, which may produce other differences.

Try making the two models identical in every way except specifying `uncertainty=True` for one and `False` for the other.  There will still be some difference, because uncertainty creates a more complicated model with a different loss function.  But it should at least be more comparable.

**Comment by amandajean119** [comment]
This was done as well:
'''
## with uncertainty
model = dc.models.MultitaskRegressor(len(tasks), 1024, uncertainty=True)
model.fit(train_dataset, nb_epoch=200)
y_pred, y_std = model.predict_uncertainty(test_dataset)
metric = dc.metrics.Metric(dc.metrics.pearson_r2_score, np.mean, mode="regression")
train_scores = model.evaluate(train_dataset, [metric], transformers)
valid_scores = model.evaluate(valid_dataset, [metric], transformers)
test_scores = model.evaluate(test_dataset, [metric], transformers)

# without uncertainty
model1 = dc.models.MultitaskRegressor(len(tasks), 1024, uncertainty=False)
model1.fit(train_dataset, nb_epoch=200)
y_pred1= model1.predict(test_dataset)
metric = dc.metrics.Metric(dc.metrics.pearson_r2_score, np.mean, mode="regression")
train_scores1 = model1.evaluate(train_dataset, [metric], transformers)
valid_scores1 = model1.evaluate(valid_dataset, [metric], transformers)
test_scores1 = model1.evaluate(valid_dataset, [metric], transformers)
'''
The data was pre-split randomly and the same train/valid/test was used on deepchem Graph convolutions models with 250 epochs. The performance comparison is shown below -

With uncertainty - train_scores : 0.803/ test_scores : 0.811/ valid_scores : 0.794
Without uncertainty - train_scores1 : 0.992/ test_scores1 : 0.865/ valid_scores1 : 0.859

I believe the table in the original post was done in the same way; I am checking with my intern Sharmi to be sure, as she is the one who ran these tests.

**Comment by peastman** [comment]
What if you keep training it longer?  Is it possible the performance just hasn't finished improving after 200 epochs?  You might also need to use a different learning rate.  It is a different model after all, so you might need to train it a bit differently.

**Comment by BSharmi** [comment]
I tried that option - training it longer unto 350 epochs. But increasing the number of epochs did not improve the performance with uncertainty = true. I tried a couple other options such as increasing number of hidden layers, using exponential decay. I did not see performance improvement with increasing the number of hidden layers. However, I noticed that with with uncertainty set to True and exponential decay, the performance improved (I tried multiple runs to reproduce the result and found the improvement consistent) with the same number of epochs as the model without uncertainty (version 2.0) and the training Pearson R2 score was increased to 0.92 if I remember correctly.
